handshake.go

```diff
@@ -174,30 +173,21 @@ func initiateHandshake(conn net.Conn, theirKey ed25519.PublicKey, ourSettings co
 }
 
 func acceptHandshake(conn net.Conn, ourKey ed25519.PrivateKey, ourSettings connSettings) (*seqCipher, connSettings, error) {
-	// exchange versions
-	var theirVersion [1]byte
-	if _, err := io.ReadFull(conn, theirVersion[:]); err != nil {
-		return nil, connSettings{}, fmt.Errorf("could not read peer version: %w", err)
-	} else if _, err := conn.Write([]byte{3}); err != nil {
-		return nil, connSettings{}, fmt.Errorf("could not write our version: %w", err)
-	} else if theirVersion[0] == 0 {
-		return nil, connSettings{}, errors.New("peer sent invalid version")
-	}
-	if theirVersion[0] < 3 {
-		return nil, connSettings{}, errors.New("versions 1 and 2 are no longer supported")
-	}
-
 	xsk, xpk := generateX25519KeyPair()
 
-	// read pubkey
-	buf := make([]byte, 32+64+connSettingsSize+chachaPoly1305TagSize)
-	if _, err := io.ReadFull(conn, buf[:32]); err != nil {
+	// read version and pubkey
+	buf := make([]byte, 1+32+64+connSettingsSize+chachaPoly1305TagSize)
+	if _, err := io.ReadFull(conn, buf[:33]); err != nil {
```

Comment:
I don't think this preserves forward compatibility. Say we change the handshake in v4. If we exchange just the version byte first, then the v4 mux will see that it's talking to v3, and send a v3 handshake. But if the v3 mux expects to read the full handshake first before sending its own version+handshake, then the v4 mux won't know which handshake to send.